File: pd_tuner_starter.py
#!/usr/bin/env python
import rospy
import time
from sensor_msgs.msg import Imu
from crazyflie_driver.srv import UpdateParams
from std_srvs.srv import Empty

class PIDTuner():
    
    def imuSubscriber(self,imu):
        x=0
        #Store Imu data
	
    def __init__(self):
        rospy.init_node('pid_tuner', anonymous=False)
        rospy.on_shutdown(self.shutdown)
        rospy.Subscriber('/crazyflie/imu', Imu, self.imuSubscriber)
        r = rospy.Rate(200)
        #################################################################
	#Initialize PID Gains for Pitch
        rospy.wait_for_service('/crazyflie/update_params')
        rospy.loginfo("found update_params service")
        self.update_params = rospy.ServiceProxy('/crazyflie/update_params', UpdateParams)#Proxy for pushing parameters to crazyflie server
	time.sleep(1)
	#Wait 1 second for the server to catch up
        rospy.set_param("/crazyflie/pid_rate/pitch_kp", 1)
        self.update_params(["pid_rate/pitch_kp"])
	rospy.set_param("/crazyflie/pid_rate/pitch_kd", 0)
        self.update_params(["pid_rate/pitch_kd"])
	rospy.set_param("/crazyflie/pid_rate/pitch_ki", 0)
        self.update_params(["pid_rate/pitch_ki"])

        while (not rospy.is_shutdown()):   
	    
	    rospy.set_param("/crazyflie/pid_rate/pitch_kp", current_kp)
            self.update_params(["pid_rate/pitch_kp"])
	    
	    #IF Oscillating update kd
	    rospy.set_param("/crazyflie/pid_rate/pitch_kd", current_kd)
            self.update_params(["pid_rate/pitch_kd"])
	    time.sleep(2)#wait 1-2 seconds for sampling Imu data
	    
	    #Main Logic Goes here
		#Calculate errors and variation
		#Update proportional and/or derivative gains

   

    def shutdown(self):
        rospy.loginfo("Stopping the pid_tuner...")
        rospy.sleep(1)
 
if __name__ == '__main__':
    try:
        PIDTuner()
	rospy.spin()
    except:
        rospy.loginfo("pid_tuner node terminated.")

File: Starter Code/Crazyflie/pd_tuner_starter.cpp
#include "ros/ros.h"
#include "sensor_msgs/Imu.h"
#include "crazyflie_driver/UpdateParams.h"
#include "std_srvs/Empty.h"
#include <string>
using namespace std;

void imuCallback(const sensor_msgs::Imu::ConstPtr& imuData)
{
  //Store Imu Data
  
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "pid_tuner");
  ros::NodeHandle nh;
  ros::Publisher vel_pub;
  ros::Subscriber sub = nh.subscribe("/crazyflie/imu", 10, imuCallback);
  ros::Rate loop_rate(200);
  float kp =1;
  float kd =0;
  //Initialize the gains
  ros::ServiceClient client = nh.serviceClient<crazyflie_driver::UpdateParams>("/crazyflie/update_params");
  crazyflie_driver::UpdateParams srv;
  std::vector<std::basic_string<char> > s;
  s.push_back("pid_rate/pitch_kp");
  s.push_back("pid_rate/pitch_kd");
  s.push_back("pid_rate/pitch_ki");
  srv.request.params =s;
  ROS_INFO("Update parameters");
  nh.setParam("/crazyflie/pid_rate/pitch_kp",1);//setting kp gain for pitch controller, update kp if the system is not oscillating yet
  nh.setParam("/crazyflie/pid_rate/pitch_kd",0);//setting kd gain for pitch controller, update kd if oscillating
  nh.setParam("/crazyflie/pid_rate/pitch_ki",0);//setting kd gain for pitch controller, update kd if oscillating
  client.call(srv);
  //Main Loop
  while (ros::ok())
  {
    ROS_INFO("Update parameters");
    nh.setParam("/crazyflie/pid_rate/pitch_kp",kp);//setting kp gain for pitch controller, update kp if the system is not oscillating yet
    nh.setParam("/crazyflie/pid_rate/pitch_kd",kd);//setting kd gain for pitch controller, update kd if oscillating
    client.call(srv);
    //wait 1-2 seconds here to collect imu data
    //Main Logic goes here, calculate errors, update gains
    ros::spinOnce();
    loop_rate.sleep();
  }
  return 0;
}

